src/Chainweb/Pact/Backend/ModuleCache.hs

```diff
@@ -0,0 +1,184 @@
+import Data.ByteString (ByteString)
+import Pact.Types.Persistence
+import Control.Lens
+import Data.Aeson (decodeStrict')
+import qualified Crypto.Hash as C (hash)
+import Crypto.Hash.Algorithms
+import qualified Data.ByteArray as BA
+import qualified Data.HashMap.Strict as HM
+import Data.Hashable
+import Data.Int
+import Pact.Types.SizeOf
+import Data.List (sort)
+
+-- -------------------------------------------------------------------------- --
+-- CacheAddress
+
+newtype CacheAddress = CacheAddress ByteString
+    deriving (Show, Eq,Ord,Hashable)
+
+-- -------------------------------------------------------------------------- --
+-- CacheEntry
+
+data CacheEntry = CacheEntry
+  { _ceTxId :: !TxId
+    -- ^ Priority for cache eviction
+  , _ceAddy :: !CacheAddress
+    -- ^ Key. Uniquly identifieds entries in the cache
+  , _ceSize :: !Int64
+    -- ^ The size of the entry. Use to keep track of the cache limit
+  , _ceData :: !PersistModuleData
+    -- ^ Cached data.
+  }
+  deriving (Show)
+
+ceTxId :: Lens' CacheEntry TxId
+ceTxId = lens _ceTxId (\a b -> a { _ceTxId = b })
+
+instance Eq CacheEntry where
+  a == b = _ceAddy a == _ceAddy b && _ceTxId a == _ceTxId b
+
+-- | Sort first by '_ceTxId', so that smaller tx ids are evicted first from the
+-- cache.
+--
+instance Ord CacheEntry where
+  a `compare` b = cmp _ceTxId (cmp _ceAddy EQ)
+    where
+      cmp acc next = case acc a `compare` acc b of
+        EQ -> next
+        c -> c
+
+-- -------------------------------------------------------------------------- --
+-- ModuleCache
+
+data ModuleCache = ModuleCache
+  { _mcStore :: !(HM.HashMap CacheAddress CacheEntry)
+  , _mcSize :: !Int64
+  , _mcLimit :: !Int64
+  }
+  deriving (Show)
+
+makeLenses 'ModuleCache
+
+emptyCache :: Int64 -> ModuleCache
+emptyCache limit = ModuleCache
+    { _mcStore = mempty
+    , _mcSize = 0
+    , _mcLimit = limit
+    }
+
+isEmpty :: ModuleCache -> Bool
+isEmpty = HM.null . _mcStore
+
+-- | The totall size of all modules in the cache.
+--
+-- Complexity: \(O(1)\)
+--
+size :: ModuleCache -> Int
+size = fromIntegral . _mcSize
+
+-- | Number of modules in the cache.
+--
+-- Complexity: \(O(n)\), linear in the number of entries
+--
+count :: ModuleCache -> Int
+count = HM.size . _mcStore
+
+-- | Decode Module from row data using the module cache.
+--
+checkModuleCache
+    :: Utf8
+        -- ^ Row Key for the module
+    -> ByteString
+        -- ^ row data that contains the encoded module
+    -> TxId
+        -- ^ TxId. This is used as priority for cache eviction. Smaller values
+        -- are evicted first.
+    -> ModuleCache
+    -> (Maybe PersistModuleData, ModuleCache)
+checkModuleCache key rowdata txid = runState $ do
+    readModuleCache txid addy >>= \case
+
+        -- Cache hit
+        Just !x -> return $ Just x
+
+        -- Cache miss: decode module and insert into cache
+        Nothing -> case decodeStrict' rowdata of
+            Nothing -> return Nothing
+            Just !m -> do
```

Comment:
Should this raise an error instead of returning `Nothing`?

```suggestion
        Nothing -> case eitherDecodeStrict' rowdata of
            e@Left{} -> return e
            Right !m -> do
```

Reply:
See response above